src/vip_ipykernel_launcher.py

```diff
@@ -0,0 +1,26 @@
+"""Taken and modified from
+https://github.com/ipython/ipykernel/blob/master/ipykernel_launcher.py
+
+Entry point for launching an ViP-IPython kernel.
+"""
+
+import subprocess
+import sys
+
+from vip_ipykernel import venv_search
+
+if __name__ == '__main__':
+    # Remove the CWD from sys.path while we load stuff.
+    # This is added back by InteractiveShellApp.init_path()
+    if sys.path[0] == '':
+        del sys.path[0]
+
+    args = sys.argv
+
+    args[0] = str(venv_search())
+
+    #  TODO: I want to use the jupyter logger to print this off but can't figure
+    #  out how to do that, @takluyver can you help with this?
+    print(f"Starting venv kernel with args: {args}")
+
+    subprocess.run(args)
```

Comment:
Potentially bad advice: for something like this I might use [os.execv](https://docs.python.org/3/library/os.html#os.execv). It's a somewhat scary low-level interface, but it goes into the new program directly under the current pid, rather than creating a new child process and then waiting for it to finish.

I think we fixed things in Jupyter so it works correctly anyway (e.g. signals go to the process group rather than the individual process), so this *shouldn't* matter. It's more of a neat trick.

Reply:
Aha, didn't even know that was a thing, the current subprocess run approach works fine but it does make sense to start it under the current pid since otherwise you'll always have two 'kernels' running, with the vip-ipykernel one just... doing nothing after it starts up the actual kernel. I'll make the change now, thanks :smile:

Reply:
Hmm I'm mildly confused by the purpose of this, and the docs don't really explain it:

> The “v” variants are good when the number of parameters is variable, with the arguments being passed in a list or tuple as the args parameter. In either case, the **arguments to the child process should start with the name of the command being run**, but this is not enforced.

The first element of the arguments passed to the process are not actually passed to it, they seem to be discarded (e.g. `os.execv("/usr/bin/echo", ["1", "2"])` will echo `2` but not `1`). So I just pass the full args list instead of `args[1:]`, since element 0 is the path to the python executable, and it'll be ignored anyway.

So `os.execv(args[0], args)` works fine but I'm wondering if you know/could explain why the first argument is dropped and why it should "start with the name of the command being run"?

Reply:
I think it's just one of those mismatches between the high-level interfaces we're used to and lower-level interfaces. It's a strong convention that argv[0] is the script or executable running in that process, and higher level APIs (like bash or the subprocess module) assume that to avoid duplication. But at the low level, it's just a convention, and you can break it if you want to.

Reply:
Aha so it's just a convention thing, I was wondering if the argument went anywhere or did... something